tephra-core/src/main/java/co/cask/tephra/distributed/TransactionServiceThriftClient.java

```diff
@@ -79,67 +84,141 @@ public void close() {
   }
 
   public Transaction startLong() throws TException {
-    return TransactionConverterUtils.unwrap(client.startLong());
+    try {
+      return TransactionConverterUtils.unwrap(client.startLong());
+    } catch (Exception e) {
+      isValid = false;
+      throw e;
+    }
   }
 
   public Transaction startShort() throws TException {
+    try {
       return TransactionConverterUtils.unwrap(client.startShort());
+    } catch (Exception e) {
+      isValid = false;
+      throw e;
+    }
   }
 
   public Transaction startShort(int timeout) throws TException {
+    try {
       return TransactionConverterUtils.unwrap(client.startShortTimeout(timeout));
+    } catch (Exception e) {
+      isValid = false;
+      throw e;
+    }
   }
 
   public boolean canCommit(Transaction tx, Collection<byte[]> changeIds) throws TException {
-
+    try {
       return client.canCommitTx(TransactionConverterUtils.wrap(tx),
                                 ImmutableSet.copyOf(Iterables.transform(changeIds, BYTES_WRAPPER))).isValue();
+    } catch (Exception e) {
+      isValid = false;
+      throw e;
+    }
   }
 
   public boolean commit(Transaction tx) throws TException {
-
+    try {
       return client.commitTx(TransactionConverterUtils.wrap(tx)).isValue();
+    } catch (Exception e) {
```

Comment:
I updated `TransactionServiceThriftClient` to catch only `TException` to mark the underlying `Client` as invalid. Additionally, I made the conversion from the thrift version of the service exceptions to the non-thrift version in the same (`TransactionServiceThriftClient`) class.
This makes it so that the client isn't marked invalid upon the following `TException`s:
`TInvalidTruncateTimeException`
`TTransactionCouldNotTakeSnapshotException`
`TTransactionNotInProgressException`